### PR TITLE
 Make tempStorageDirectory configuration optional and rely on task dir instead

### PR DESCRIPTION
**Pull request by adarshsanjeev**
Currently, durable storage and export both require configuring a temporary directory to be used using `druid.export.storage.<connectorType>.tempLocalDir` and `druid.msq.intermediate.storage.tempDir`. 

Tasks on middle manager already have a configured temporary directory. This PR aims to reduce the configuration required by using the task directory as a default if it is not explicitly configured, thus reducing the number of configs that a user has to set.

Please note that preference would be given to the user configured, `druid.*.storage.temp*Dir`, on the tasks. If that is not configured, we then use the `configured temporary directory`. 

Overlord and brokers also require storage connector configurations (for the durableStorageCleanerOverlordDuty and to fetch results of async queries respectively), but do not have a default temporary task directory. The configuration is still required for these services.

<hr>

### Release notes

`druid.export.storage.<google/s3>.tempLocalDir` and `druid.msq.intermediate.storage.tempDir` are no longer required configurations. If not configured, the task defaults to using the task temp directory.
<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by LakshSingla** [comment]
> Overlord and brokers also require storage connector configurations (for the durabelStorageCleanerOverlordDuty and to fetch results of async queries respectively), but do not have a default temporary task directory

Should Druid rely on `java.io.tmpDir` in that case? FileUtils has multiple options to create temp directories. `InputSourceSampler` in overlord uses one of those methods to create a temp directory.

File: processing/src/main/java/org/apache/druid/storage/StorageConnectorProvider.java [hunk 1]
```diff
@@ -20,9 +20,11 @@
...
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
-import com.google.inject.Provider;
+
+import java.io.File;
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
-public interface StorageConnectorProvider extends Provider<StorageConnector>
+public interface StorageConnectorProvider
 {
+  StorageConnector createStorageConnector(File tempDir);
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/storage/StorageConnectorProvider.java:29]: Could you please add java docs for this method ?
The current implementation just ignores this tempDir if the property is already set which might confuse the caller. 

Is it possible to not change this interface and have the implementation take  the "tempFIle" as the constructor param. Based on that, we can set the correct tempDir values.

**Review by cryptoe — COMMENTED** [review]
Everything else seems cool to me.

**Review by cryptoe — APPROVED** [review]

**Comment by cryptoe** [comment]
@adarshsanjeev I updated the description. Please check.

**Comment by adarshsanjeev** [comment]
@cryptoe could you please take a look at the changes I made now to MSQDurableStorageModule?

**Review by cryptoe — APPROVED** [review]

**Pull request merged**